### ISSUE TITLE
Incorrect removal of function wrapper pollutes global scope

### ISSUE THREAD

**Issue by DerGernTod** [google/closure-compiler#3144]
Closure compiler incorrectly removes a function wrapper, causing a variable to be leaked to the global scope. It's easily reproducable using ADVANCED_OPTIMIZATIONS mode with this small code snippet:

```js
(function() {
    var z = window;
    if (z.someProp) {
        z.console && z.console.log("log something");
        return;
    } 

    z.console && z.console.log("log something else");
})();
```

result: 
```js
var a=window;a.a?a.console&&a.console.log("log something"):a.console&&a.console.log("log something else");
```

this means the global scope is polluted with the `a` variable, causing it to override any previously defined `global.a` property.

**Comment by lauraharker** [comment]
This is intentional. The compiler assumes that no global variables outside of those provided in externs exist while doing optimizations.

Consider using this flag instead: `--isolation_mode=IIFE`

https://github.com/google/closure-compiler/wiki/Flags-and-Options#output

**Comment by DerGernTod** [comment]
thanks, that did the trick!

**Comment by lauraharker** [comment]
> however, i wonder why the setting is not enabled by default if i enable 'assume_function_wrapper'

Sorry to only notice this question months later. The compiler automatically turns on `assume_function_wrapper` if you use `isolation_mode=IIFE`. However, some projects don't use `isolation_mode` but still wrap their output in an IIFE. 

 For example, some projects define a custom output wrapper with `--output_wrapper`, which may or may not be an IIFE. Those projects have to explicitly enable `assume_function_wrapper`.
